src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -49,57 +51,61 @@ public class Regex
 
         protected internal RegexRunnerFactory factory;
 
-        internal Dictionary<int, int> _caps;            // if captures are sparse, this is the hashtable capnum->index
-        internal Dictionary<string, int> _capnames;     // if named captures are used, this maps names->index
+        protected internal Hashtable caps;            // if captures are sparse, this is the hashtable capnum->index
+        protected internal Hashtable capnames;     // if named captures are used, this maps names->index
 
-        protected internal String[] capslist;              // if captures are sparse or named captures are used, this is the sorted list of names
+        protected internal string[] capslist;              // if captures are sparse or named captures are used, this is the sorted list of names
         protected internal int capsize;                    // the size of the capture array
+
+        [CLSCompliant(false)]
         protected IDictionary Caps
         {
             get
             {
-                return _caps;
+                return caps;
             }
             set
             {
                 if (value == null)
                     throw new ArgumentNullException(nameof(value));
-
-                _caps = value as Dictionary<int, int>;
-                if (_caps == null)
+                
+                caps = value as Hashtable;
+                if (caps == null)
                 {
-                    _caps = new Dictionary<int, int>(value.Count);
+                    caps = new Hashtable();
```

Comment:
> caps = new Hashtable(); [](start = 19, length = 24)

You lost the value.Count going to the constructor.

Or just do 
caps = new Hashtable(value)